### PR TITLE
feat: add update instance to allow changing SLA

### PR DESCRIPTION
**Pull request by mweibel**
ref: SPKS-121

### CODE REVIEW

File: pkg/crossplane/metadata.go [hunk 1]
```diff
@@ -46,6 +47,26 @@ const (
 	SLALabel = SynToolsBase + "/sla"
 )
 
+const (
+	// SLAPremium represents the string for the premium SLA
+	SLAPremium = "premium"
+	// SLAStandard is an empty string
+	SLAStandard = ""
```
Comment (tobru) [inline pkg/crossplane/metadata.go]: I'm a fan to make things explicit. Would it make sense to write "standard" into this field?

Comment (mweibel) [reply pkg/crossplane/metadata.go]: @tobru I'd prefer that as well but ATM the crossplane plans without premium SLA have the field empty - I merely reflect that fact in the code here. Not sure why it's empty and not `standard` - maybe @srueg can shed some light on this?

Comment (srueg) [reply pkg/crossplane/metadata.go]: Yes, absolutely. I'll add `standard` to the plans :+1:

Comment (srueg) [reply pkg/crossplane/metadata.go]: Done https://git.vshn.net/swisscompks/component-spks-crossplane/-/merge_requests/6

Comment (mweibel) [reply pkg/crossplane/metadata.go]: thanks :+1: 

fixed.

**Comment by mweibel** [comment]
should be ready to review :+1:

File: pkg/crossplane/instance.go [hunk 2]
```diff
@@ -151,3 +155,58 @@ func (cp *Crossplane) GetInstance(ctx context.Context, instanceID string) (*comp
...
+	} {
+		instanceLabels[l] = newPlan.Labels[l]
+	}
+	instance.SetLabels(instanceLabels)
+
+	gvk, err := gvkFromPlan(newPlan)
+	if err != nil {
+		return err
+	}
+	instance.SetGroupVersionKind(gvk)
```
Comment (srueg) [inline pkg/crossplane/instance.go:221]: This shouldn't be necessary since it's already set, no?

File: pkg/crossplane/instance.go [hunk 3]
```diff
@@ -151,3 +155,58 @@ func (cp *Crossplane) GetInstance(ctx context.Context, instanceID string) (*comp
...
+
+	newPlan, err := cp.GetPlan(ctx, planID)
+	if err != nil {
+		return err
+	}
+
+	instanceSLA := instanceLabels[SLALabel]
+	newPlanSLA := newPlan.Labels[SLALabel]
+
+	slaChangePermitted := func() bool {
```
Comment (srueg) [inline pkg/crossplane/instance.go:177]: As far as I understand, this doesn't catch the case where for example `medium` will be updated to `large-premium`?

File: pkg/crossplane/metadata.go [hunk 4]
```diff
@@ -68,5 +75,5 @@ func ConvertError(ctx context.Context, err error) error {
 		fmt.Errorf("%w (correlation-id: %q)", err, id),
 		http.StatusInternalServerError,
 		"internal-server-error",
-	).WithErrorKey(string(kErr.ErrStatus.Reason)).Build()
```
Comment (srueg) [inline pkg/crossplane/metadata.go:71]: What's the reason to remove this? I fear without the error-key it's not a valid OSB error.

**Review by srueg — CHANGES_REQUESTED** [review]

File: pkg/crossplane/instance.go (hunk 2 above)
Comment (mweibel) [reply pkg/crossplane/instance.go:221]: I thought as well but it's somehow lost - I have to dig deeper to check why.

File: pkg/crossplane/instance.go (hunk 3 above)
Comment (mweibel) [reply pkg/crossplane/instance.go:177]: exactly - wasn't specified exactly in the ticket therefore I wasn't sure if it's required. Question is how to check? string substitution is doable but not very pretty :)

File: pkg/crossplane/metadata.go (hunk 4 above)
Comment (mweibel) [reply pkg/crossplane/metadata.go:71]: `kErr` is nil at this point

File: pkg/crossplane/instance.go (hunk 3 above)
Comment (srueg) [reply pkg/crossplane/instance.go:177]: I fear it's gonna be string checking for now, yes :disappointed:

File: pkg/crossplane/instance.go (hunk 2 above)
Comment (mweibel) [reply pkg/crossplane/instance.go:221]: https://github.com/kubernetes/apimachinery/blob/master/pkg/runtime/helper.go#L250-L259

This decoder is used and removes GVK apparently. Not sure what the reasoning behind is, TBH.

File: pkg/crossplane/instance.go [hunk 5]
```diff
@@ -151,3 +155,58 @@ func (cp *Crossplane) GetInstance(ctx context.Context, instanceID string) (*comp
...
+			return true
+		}
+		if instanceSLA == SLAStandard && newPlanSLA == SLAPremium {
+			return true
+		}
+		return false
+	}
+
+	if !slaChangePermitted() {
+		return ErrSLAChangeNotPermitted
```
Comment (srueg) [inline pkg/crossplane/instance.go:203]: According to the spec, this should return a `422` error code. So we probably should return an APIError here?

https://github.com/openservicebrokerapi/servicebroker/blob/master/spec.md#response-5

Comment (mweibel) [reply pkg/crossplane/instance.go:203]: you're absolutely right.. and I had it once (not returning an APIError but handling the errors on the caller side) but I removed it for some reason. Will re-add that.

**Review by srueg — APPROVED** [review]

**Review by tobru — APPROVED** [review]

**Pull request merged**